File: frontend/src/app/core/services/setup-status.service.ts
import { HttpClient } from '@angular/common/http';
import { TeacherAssignmentService } from './teacher-assignment.service';
import { Injectable, computed, inject, signal } from '@angular/core';
import { Observable, forkJoin, map, tap } from 'rxjs';
import { environment } from '@env/environment';
import { SetupStatus, SetupStep, SetupStepKey } from '@core/models/setup.models';
import {
  CLASSROOM_DATA_SOURCE, CURRICULUM_DATA_SOURCE, ENROLLMENT_DATA_SOURCE,
  FEE_DATA_SOURCE, REFERENCE_DATA_SOURCE, TEACHER_DATA_SOURCE
} from '@core/datasource/data-source';

/** Wording of each step, in the order a school actually configures itself. */
interface StepDefinition {
  key: SetupStepKey;
  label: string;
  description: string;
  actionRoute: string;
  actionLabel: string;
}

const STEPS: readonly StepDefinition[] = [
  { key: 'ACADEMIC_YEAR', label: "Activer l'année scolaire",
    description: "Définissez la période de référence et ouvrez la campagne d'inscriptions.",
    actionRoute: '/administration', actionLabel: "Configurer l'année" },
  { key: 'CYCLES', label: 'Définir les cycles',
    description: 'Préscolaire, primaire, collège, lycée : indiquez ce que vous enseignez.',
    actionRoute: '/onboarding', actionLabel: 'Choisir les cycles' },
  { key: 'LEVELS', label: 'Créer les niveaux',
    description: "Les niveaux de chaque cycle, dans l'ordre de progression des élèves.",
    actionRoute: '/onboarding', actionLabel: 'Définir les niveaux' },
  { key: 'CLASSES', label: 'Créer les classes',
    description: 'Les classes de chaque niveau, avec leur nom et leur capacité maximale.',
    actionRoute: '/classes', actionLabel: 'Gérer les classes' },
  { key: 'SUBJECTS', label: 'Déclarer les matières',
    description: "La liste des matières enseignées dans l'établissement.",
    actionRoute: '/subjects', actionLabel: 'Gérer les matières' },
  { key: 'CURRICULUM', label: 'Programme et coefficients',
    description: 'Rattachez les matières à chaque niveau avec leur coefficient : '
      + 'sans cela, aucune moyenne ne peut être calculée.',
    actionRoute: '/subjects', actionLabel: 'Définir le programme' },
  { key: 'FEES', label: 'Plan de facturation',
    description: 'Les frais par niveau et leur échéancier, appliqués à chaque inscription.',
    actionRoute: '/finance', actionLabel: 'Définir le plan' },
  { key: 'TEACHERS', label: 'Ajouter les enseignants',
    description: "Le personnel enseignant de l'établissement.",
    actionRoute: '/teachers', actionLabel: 'Ajouter des enseignants' },
  { key: 'ASSIGNMENTS', label: 'Affecter les enseignants',
    description: 'Qui enseigne quelle matière, à quelle classe. '
      + 'Détermine aussi ce que chacun peut saisir.',
    actionRoute: '/teacher-assignments', actionLabel: 'Affecter aux classes' },
  { key: 'STUDENTS', label: 'Inscrire les élèves',
    description: 'Vos premiers élèves, un à un ou par import Excel.',
    actionRoute: '/enrollments', actionLabel: 'Inscrire des élèves' }
];

/**
 * Tracks how far the school's configuration has got.
 *
 * <p>Held in a signal so the sidebar badge, the dashboard banner and the setup
 * page all read the same value, and all refresh together after any change.</p>
 *
 * <p>The count of each step is <em>derived from the data</em>, never remembered.
 * A step done by hand, outside the wizard, ticks itself off; undo it and it
 * unticks. Demo mode obeys the same rule: it recomputes from the mock data
 * sources instead of returning a frozen snapshot, otherwise the indicator would
 * claim to read reality while showing a fixed picture.</p>
 */
@Injectable({ providedIn: 'root' })
export class SetupStatusService {
  private readonly http = inject(HttpClient);
  private readonly reference = inject(REFERENCE_DATA_SOURCE);
  private readonly curriculum = inject(CURRICULUM_DATA_SOURCE);
  private readonly classrooms = inject(CLASSROOM_DATA_SOURCE);
  private readonly teachers = inject(TEACHER_DATA_SOURCE);
  private readonly enrollments = inject(ENROLLMENT_DATA_SOURCE);
  private readonly fees = inject(FEE_DATA_SOURCE);
  private readonly assignments = inject(TeacherAssignmentService);

  private readonly _status = signal<SetupStatus | null>(null);
  readonly status = this._status.asReadonly();

  /** Drives the sidebar badge; null once everything is done. */
  readonly badge = computed(() => {
    const status = this._status();
    if (!status || status.complete) {
      return null;
    }
    return `${status.completedSteps}/${status.totalSteps}`;
  });

  readonly incomplete = computed(() => {
    const status = this._status();
    return status !== null && !status.complete;
  });

  load(): Observable<SetupStatus> {
    const source = environment.useMockData
      ? this.mockStatus()
      : this.http.get<SetupStatus>(`${environment.apiBaseUrl}/school/setup-status`);

    return source.pipe(tap((status) => this._status.set(status)));
  }

  /** Called after any screen that could complete a step. */
  refresh(): void {
    this.load().subscribe({ error: () => undefined });
  }

  /**
   * Demo mode: same ten steps, counted from the mock data sources.
   *
   * Assignment counts come from the same board used by the assignment screen.
   */
  private mockStatus(): Observable<SetupStatus> {
    return forkJoin({
      years: this.reference.academicYears(),
      assignments: this.assignments.board(),
      levels: this.curriculum.levels(),
      subjects: this.curriculum.listSubjects(),
      classes: this.classrooms.list(),
      teachers: this.teachers.search({ page: 0, size: 200 }),
      enrollments: this.enrollments.search({ page: 0, size: 1 }),
      fees: this.fees.levels()
    }).pipe(map((data) => {
      const cycles = new Set(data.levels.map((l) => l.cycleId));
      const assigned = data.assignments.assignments.length;

      const counts: Record<SetupStepKey, number> = {
        ACADEMIC_YEAR: data.years.filter((y) => y.status === 'ACTIVE').length,
        CYCLES: cycles.size,
        LEVELS: data.levels.length,
        CLASSES: data.classes.length,
        SUBJECTS: data.subjects.length,
        // Comme sur le serveur : un niveau ne compte que s'il porte au moins
        // une matière notée. Ouvrir un programme vide ne coche rien.
        CURRICULUM: data.levels.filter((l) => l.ready).length,
        // Un niveau ne compte que s'il porte au moins un frais obligatoire :
        // un tarif facultatif seul ne facture rien à l'inscription.
        FEES: data.fees.filter((l) => l.ready).length,
        TEACHERS: data.teachers.totalElements,
        ASSIGNMENTS: assigned,
        STUDENTS: data.enrollments.totalElements
      };

      return this.assemble(counts);
    }));
  }

  private assemble(counts: Record<SetupStepKey, number>): SetupStatus {
    const steps: SetupStep[] = STEPS.map((definition) => ({
      key: definition.key,
      label: definition.label,
      description: definition.description,
      done: counts[definition.key] > 0,
      required: true,
      count: counts[definition.key],
      actionRoute: definition.actionRoute,
      actionLabel: definition.actionLabel
    }));

    const done = steps.filter((s) => s.done).length;
    return {
      schoolId: 'demo',
      schoolName: environment.schoolName,
      academicYearCode: '2026-2027',
      completedSteps: done,
      totalSteps: steps.length,
      percentComplete: Math.round((done * 100) / steps.length),
      complete: done === steps.length,
      nextStepKey: steps.find((s) => !s.done)?.key,
      steps
    };
  }
}

File: backend/src/main/java/ci/company/eduops/school/service/SetupStatusService.java
package ci.company.eduops.school.service;

import ci.company.eduops.academicyear.domain.AcademicYear;
import ci.company.eduops.academicyear.domain.AcademicYearStatus;
import ci.company.eduops.academicyear.repository.AcademicYearRepository;
import ci.company.eduops.classroom.repository.ClassroomRepository;
import ci.company.eduops.common.exception.BusinessException;
import ci.company.eduops.common.exception.ErrorCode;
import ci.company.eduops.common.tenant.TenantContext;
import ci.company.eduops.curriculum.repository.CurriculumRepository;
import ci.company.eduops.curriculum.repository.TeacherAssignmentRepository;
import ci.company.eduops.cycle.repository.CycleRepository;
import ci.company.eduops.enrollment.repository.EnrollmentRepository;
import ci.company.eduops.finance.repository.FeeScheduleRepository;
import ci.company.eduops.level.repository.LevelRepository;
import ci.company.eduops.school.domain.School;
import ci.company.eduops.school.dto.response.SetupStatusResponse;
import ci.company.eduops.school.dto.response.SetupStepResponse;
import ci.company.eduops.school.repository.SchoolRepository;
import ci.company.eduops.subject.repository.SubjectRepository;
import ci.company.eduops.teacher.repository.TeacherRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Works out how far a school has got with its configuration.
 *
 * <p>Deliberately derived from the data rather than stored as a flag. An
 * administrator who creates classes by hand, imports pupils from Excel, or
 * abandons the wizard halfway still gets an accurate picture — and nothing has
 * to be kept in sync.</p>
 */
@Service
public class SetupStatusService {

    private final SchoolRepository schoolRepository;
    private final AcademicYearRepository academicYearRepository;
    private final CycleRepository cycleRepository;
    private final LevelRepository levelRepository;
    private final ClassroomRepository classroomRepository;
    private final SubjectRepository subjectRepository;
    private final FeeScheduleRepository feeScheduleRepository;
    private final TeacherRepository teacherRepository;
    private final EnrollmentRepository enrollmentRepository;
    private final CurriculumRepository curriculumRepository;
    private final TeacherAssignmentRepository assignmentRepository;

    public SetupStatusService(SchoolRepository schoolRepository,
                              AcademicYearRepository academicYearRepository,
                              CycleRepository cycleRepository,
                              LevelRepository levelRepository,
                              ClassroomRepository classroomRepository,
                              SubjectRepository subjectRepository,
                              FeeScheduleRepository feeScheduleRepository,
                              TeacherRepository teacherRepository,
                              EnrollmentRepository enrollmentRepository,
                              CurriculumRepository curriculumRepository,
                              TeacherAssignmentRepository assignmentRepository) {
        this.schoolRepository = schoolRepository;
        this.academicYearRepository = academicYearRepository;
        this.cycleRepository = cycleRepository;
        this.levelRepository = levelRepository;
        this.classroomRepository = classroomRepository;
        this.subjectRepository = subjectRepository;
        this.feeScheduleRepository = feeScheduleRepository;
        this.teacherRepository = teacherRepository;
        this.enrollmentRepository = enrollmentRepository;
        this.curriculumRepository = curriculumRepository;
        this.assignmentRepository = assignmentRepository;
    }

    @Transactional(readOnly = true)
    public SetupStatusResponse currentStatus() {
        UUID schoolId = TenantContext.getSchoolId();
        if (schoolId == null) {
            throw BusinessException.of(ErrorCode.PORTAL_PROFILE_MISSING,
                    "Aucun établissement associé à ce compte.");
        }

        School school = schoolRepository.findById(schoolId)
                .orElseThrow(() -> BusinessException.of(ErrorCode.SCHOOL_NOT_FOUND));

        AcademicYear year = academicYearRepository
                .findBySchoolIdAndStatus(schoolId, AcademicYearStatus.ACTIVE)
                .orElse(null);
        UUID yearId = year == null ? null : year.getId();

        List<SetupStepResponse> steps = new ArrayList<>();

        // Ordered the way a school actually sets itself up: nothing below can
        // be done before the line above it exists.
        steps.add(new SetupStepResponse(
                "ACADEMIC_YEAR", "Activer l'année scolaire",
                "Définissez la période de référence et ouvrez la campagne d'inscriptions.",
                true,
                year == null ? 0 : 1,
                "/administration",
                "Configurer l'année"));

        steps.add(new SetupStepResponse(
                "CYCLES", "Définir les cycles",
                "Préscolaire, primaire, collège, lycée : indiquez ce que vous enseignez.",
                true,
                cycleRepository.findBySchoolIdOrderBySequenceAsc(schoolId).size(),
                "/onboarding",
                "Choisir les cycles"));

        steps.add(new SetupStepResponse(
                "LEVELS", "Créer les niveaux",
                "Les niveaux de chaque cycle, dans l'ordre de progression des élèves.",
                true,
                countLevels(schoolId),
                "/onboarding",
                "Définir les niveaux"));

        steps.add(new SetupStepResponse(
                "CLASSES", "Créer les classes",
                "Les classes de chaque niveau, avec leur nom et leur capacité maximale.",
                true,
                yearId == null ? 0 : classroomRepository.countActive(yearId),
                "/classes",
                "Gérer les classes"));

        steps.add(new SetupStepResponse(
                "SUBJECTS", "Déclarer les matières",
                "La liste des matières enseignees dans l'établissement.",
                true,
                subjectRepository.findBySchoolIdAndStatusOrderByNameAsc(
                        schoolId, ci.company.eduops.common.domain.CommonStatus.ACTIVE).size(),
                "/subjects",
                "Gérer les matières"));

        steps.add(new SetupStepResponse(
                "CURRICULUM", "Programme et coefficients",
                "Rattachez les matières à chaque niveau avec leur coefficient : sans cela, aucune moyenne ne peut être calculée.",
                true,
                yearId == null ? 0 : curriculumRepository.countReady(yearId),
                "/subjects",
                "Définir le programme"));

        steps.add(new SetupStepResponse(
                "FEES", "Plan de facturation",
                "Les frais par niveau et leur échéancier, appliqués à chaque inscription.",
                true,
                yearId == null ? 0 : feeScheduleRepository.countPricedLevels(yearId),
                "/finance",
                "Définir le plan"));

        steps.add(new SetupStepResponse(
                "TEACHERS", "Ajouter les enseignants",
                "Le personnel enseignant de l'établissement.",
                true,
                teacherRepository.countBySchoolIdAndStatus(
                        schoolId, ci.company.eduops.teacher.domain.TeacherStatus.ACTIVE),
                "/teachers",
                "Ajouter des enseignants"));

        steps.add(new SetupStepResponse(
                "ASSIGNMENTS", "Affecter les enseignants",
                "Qui enseigne quelle matière, à quelle classe. Détermine aussi ce que chacun peut saisir.",
                true,
                yearId == null ? 0 : assignmentRepository.countActiveForYear(yearId),
                "/teacher-assignments",
                "Affecter aux classes"));

        steps.add(new SetupStepResponse(
                "STUDENTS", "Inscrire les élèves",
                "Vos premiers élèves, un à un ou par import Excel.",
                true,
                yearId == null ? 0 : enrollmentRepository.countActiveForYear(yearId),
                "/enrollments",
                "Inscrire des élèves"));

        return summarise(school, year, steps);
    }

    /** Levels across every cycle of the school. */
    private long countLevels(UUID schoolId) {
        return cycleRepository.findBySchoolIdOrderBySequenceAsc(schoolId).stream()
                .mapToLong(cycle -> levelRepository
                        .findByCycleIdOrderBySequenceAsc(cycle.getId()).size())
                .sum();
    }

    private SetupStatusResponse summarise(School school, AcademicYear year,
                                          List<SetupStepResponse> steps) {
        SetupStatusResponse response = new SetupStatusResponse();
        response.setSchoolId(school.getId());
        response.setSchoolName(school.getName());
        response.setAcademicYearCode(year == null ? null : year.getCode());
        response.setSteps(steps);

        List<SetupStepResponse> required = steps.stream()
                .filter(SetupStepResponse::isRequired)
                .toList();

        int done = (int) required.stream().filter(SetupStepResponse::isDone).count();
        response.setTotalSteps(required.size());
        response.setCompletedSteps(done);
        response.setPercentComplete(required.isEmpty()
                ? 100
                : Math.round(done * 100f / required.size()));
        response.setComplete(done == required.size());

        response.setNextStepKey(required.stream()
                .filter(step -> !step.isDone())
                .map(SetupStepResponse::getKey)
                .findFirst()
                .orElse(null));

        return response;
    }
}
